[MRG + 2] extending tfidf documentation

rasbt opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes #6996
#### What does this implement/fix? Explain your changes.

Adds the Tf-idf equation to the `TfidfTransformer`'s docstring and adds equation + hands-on example to the user guide documentation.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

rasbt committed:
extending tfidf documentation

nelson-liu commented on doc/modules/feature_extraction.rst:
```diff
@@ -398,18 +398,34 @@ suitable for usage by a classifier it is very common to use the tf–idf
+
+Please note that the tf-idfs computed in scikit-learn's 
+:class:`TfidfTransformer` and :class:`TfidfVectorizer` differ slightly from the
+standard textbook notation, which is typically defines the idf as as
```
This was a bit ambiguous to me upon first reading it, i think it'd be more clear to say "differ slightly from the standard textbook notation that defines the idf as:", since i wasn't sure initially if you were referring to the textbook notation or the scikit-learn notation.

If you think it's ok the way it is now, can you at least remove the "is" and the duplicated "as"?

amueller commented on sklearn/feature_extraction/text.py:
```diff
@@ -926,11 +926,21 @@ class TfidfTransformer(BaseEstimator, TransformerMixin):
-    settings that correspond to the SMART notation used in IR, as follows:
+    The formula that is used to compute the tf-idf of term t is
+    tf-idf(d, t) = tf(t) * idf(d, t), and the idf is computed as
+    idf(d, t) = log [ n / df(d, t) ] + 1,
```
I thought that there were a bunch more "+1" by default, like in both the numerator and denominator.

nelson-liu replied:
yup, the default parameter of `smooth_idf` is `True`, so the idf formula is `idf(d, t) = log [ (1+n) / (1+df(d, t)) ] + 1` by default.

nelson-liu commented on doc/modules/feature_extraction.rst:
```diff
@@ -433,21 +449,83 @@ content of the documents::
+:math:`\text{idf}(d, t)_{\text{term1}} = 
+log \frac{n_d}{\text{df}(d, t)} + 1 = log(1)+1 = 1`
+
+:math:`\text{tf-idf}_{\text{term1}} = \text{tf} \times \text{idf} = 3 \times 1 = 3`
```
this example is very nice :)

rasbt committed:
polish and clarify tfidf description

rasbt commented:
@nelson-liu 
> This was a bit ambiguous to me upon first reading it, i think it'd be more clear to say "differ slightly from the standard textbook notation that defines the idf as:", since i wasn't sure initially if you were referring to the textbook notation or the scikit-learn notation.

> If you think it's ok the way it is now, can you at least remove the "is" and the duplicated "as"?

Thanks Nelson, I agree! Just changed it!


@amueller 
> I thought that there were a bunch more "+1" by default, like in both the numerator and denominator.

Yeah, like Nelson mentioned, there are two "+1" (1 in the denominator and one in the numerator) if `smooth_idf` is set to `True` (I thought I would explain it incrementally, but I added more notes about that.

rasbt committed:
fix doctest

rasbt committed:
fix doctest part 2

amueller commented on doc/modules/feature_extraction.rst:
```diff
@@ -398,18 +398,34 @@ suitable for usage by a classifier it is very common to use the tf–idf
+In the :class:`TfidfTransformer` and :class:`TfidfVectorizer`, the
+"1" count is added to the idf instead of the idf's denominator:
+
+:math:`\text{idf}(t) = log{\frac{n_d}{\text{df}(d,t)}} + 1`
```
this is not the full formula with smooth_idf=True, right?

rasbt commented:
@amueller  

> this is not the full formula with smooth_idf=True, right?

Correct, this is *not* the full formula with smooth_idf=True. I wanted to explain this gradually, since I thought that it would be "too much" for a reader otherwise (and you would have to explain that you have to take the "+1"s away if you set smooth_idf=False. So, I introduced the "raw" tf-idfs first, and after the example, I get to the smooth_idf:

> Furthermore, the default parameter smooth_idf=True adds "1" to the numerator and denominator as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions:  
 \text{idf}(t) = log{\frac{1 + n_d}{1+\text{df}(d,t)}} + 1  
Using this modification, the tf-idf of the third term in document 1 changes to 1.8473:

amueller commented:
I'm not sure this is a good approach. Skimming the docs, it's really hard to pick out which is the formula that's actually used. You need to read a pretty big chunk of text before you can actually find out.

rasbt commented:
> You need to read a pretty big chunk of text before you can actually find out.

Hm, yeah, you are right, but I think that it would make it easier for people to follow the rationale behind the scikit-learn implementation (since it is different from the classic implementations in literature). Not sure about this, I think it helps new people to understand what's going on, but it could be too much for people who just quickly want to glance at the equation. So, how about adding the equation with  `smooth_idf=True` at the beginning for a quick glance. Then, we could add the tf-idf example to explain it further? Something like

"...the tf-idf is computed as [insert equation] , where `smooth_idf=True` (the default). To explain the rationale behind this equation, let's take a look at the  ...

On the other hand, that's the "Tutorial" on text classification, and we have a shorter version in the docstring; I think that a verbose version somewhere is not such a bad idea so that people can follow along what the Tf-idf is doing exactly ...

nelson-liu commented:
>On the other hand, that's the "Tutorial" on text classification, and we have a shorter version in the docstring; I think that a verbose version somewhere is not such a bad idea so that people can follow along what the Tf-idf is doing exactly ...

Yeah, I agree that the verbose version here is a good idea because it's quite confusing to figure out what is happening in our implementation of tf-idf without it. if anything, putting the full equation in the docstring would be the way to go imo (since i'd look there for that info before looking at the tutorial)

amueller commented:
> So, how about adding the equation with smooth_idf=True at the beginning for a quick glance. Then, we culd add the tf-idf example to explain it further?

+1

rasbt committed:
present default-setting equation before examples

rasbt commented:
Okay, I added a the `smooth_idf` equation for the idf at the top now, just before the example section. I actually liked it better before, but yeah, let me know what you think :P

amueller commented:
I like it :P 
LGTM.

rasbt commented:
Is there sth that needs to be done regarding circleci or is this okay without rebase (looks like I haven't rebased the "what's new" file and it is missing some reference or so).

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -926,11 +926,26 @@ class TfidfTransformer(BaseEstimator, TransformerMixin):
-    settings that correspond to the SMART notation used in IR, as follows:
+    The formula that is used to compute the tf-idf of term t is
+    tf-idf(d, t) = tf(t) * idf(d, t), and the idf is computed as
+    idf(d, t) = log [ n / df(d, t) ] + 1 (if `smooth_idf=False`),
```
Single backtick does nothing. And I wonder if the rest of this should be formatted.

jnothman commented:
Apart from that formatting, LGTM

rasbt commented:
@jnothman 

Ah, right, that's a habit of mine since I typically use a docstring to markdown converter. I do not have a strong preference for double-back ticks vs no back ticks. Glancing over some other docstrings, I think there's no clear consistency either. I would just remove the backticks then for simplicity then if that's okay?

nelson-liu commented:
hmm, i'm still partial to 2 backticks myself; parameter values in code-blocks make them easier to read / distinguish.

jnothman commented:
For things that are actually public identifiers (parameter names and attributes) I much prefer using tt format. I know it's not very consistent -- apart from anything else the parameter and attribute listing in numpydoc and function signature in sphinx's autodoc fail to use tt -- but I think it's worth doing when we remember. It's most useful in narrative strings where the meaning is ambiguous and when the eye needs to be drawn to important terms.

rasbt commented:
Alright then, I added some backticks :)

rasbt committed:
try to fix the merge conflict part 2

jnothman commented:
I mean if we wanted to go all out, we'd have invented a sphinx extension to link to parameter and attribute documentation and format appropriately... but we haven't :)

jnothman commented:
Thanks @rasbt, the quality of your contributions is much appreciated.

amueller commented:
thanks @rasbt !

jnothman commented:
See the `smooth_idf` parameter
